server/ec2spotmanager/tasks.py

```diff
@@ -1 +1,562 @@
+                    fulfilled.add_tag(SPOTMGR_TAG + "-Updatable", "1")
+
+            if expired:
+                logger.info("[Pool %d] %d spot requests timed out, cancelling.", pool.id, len(expired))
+                cluster.cancel_spot_requests(expired)
+                for instance_id in expired:
+                    instances_by_ids[instance_id].delete()
+
+            boto_instances = cluster.find(filters={"tag:" + SPOTMGR_TAG + "-PoolId": str(pool.pk)})
+
+            for boto_instance in boto_instances:
+                # Store ID seen for debugging purposes
+                debug_boto_instance_ids_seen.add(boto_instance.id)
+
+                # state_code is a 16-bit value where the high byte is
+                # an opaque internal value and should be ignored.
+                state_code = boto_instance.state_code & 255
+
+                if (SPOTMGR_TAG + "-Updatable" not in boto_instance.tags or
+                        int(boto_instance.tags[SPOTMGR_TAG + "-Updatable"]) <= 0):
+                    # The instance is not marked as updatable. We must not touch it because
+                    # a spawning thread is still managing this instance. However, we must also
+                    # remove this instance from the instances_left list if it's already in our
+                    # database, because otherwise our code here would delete it from the database.
+                    if boto_instance.id in instance_ids_by_region[region]:
+                        instances_left.remove(instances_by_ids[boto_instance.id])
+                    else:
+                        debug_not_updatable_continue.add(boto_instance.id)
+                    continue
+
+                instance = None
+
+                # Whenever we see an instance that is not in our instance list for that region,
+                # make sure it's a terminated instance because we should never have a running
+                # instance that matches the search above but is not in our database.
+                if boto_instance.id not in instance_ids_by_region[region]:
+                    if state_code not in [INSTANCE_STATE['shutting-down'], INSTANCE_STATE['terminated']]:
+
+                        # As a last resort, try to find the instance in our database.
+                        # If the instance was saved to our database between the entrance
+                        # to this function and the search query sent to EC2, then the instance
+                        # will not be in our instances list but returned by EC2. In this
+                        # case, we try to load it directly from the database.
+                        q = Instance.objects.filter(ec2_instance_id=boto_instance.id)
+                        if q:
+                            instance = q[0]
+                            logger.error("[Pool %d] Instance with EC2 ID %s was reloaded from database.",
```

Comment:
We should check where these go now that it runs as a task (and possibly keep it in its own log if that is doable).